src/hotspot/share/opto/memnode.cpp

```diff
@@ -1269,6 +1267,59 @@ bool LoadNode::is_instance_field_load_with_local_phi(Node* ctrl) {
   return false;
 }
 
+bool LoadNode::is_boxed_value_load_with_local_phi(PhaseGVN* phase) {
+  Compile* C = phase->C;
+  Node* address = in(Address);
+  const TypeOopPtr *t_oop = address->bottom_type()->isa_oopptr();
+  intptr_t ignore = 0;
+  Node*    base = AddPNode::Ideal_base_and_offset(address, phase, ignore);
+  bool base_is_phi = (base != NULL) && base->is_Phi();
+  bool load_boxed_value = t_oop != NULL && t_oop->is_ptr_to_boxed_value()
+                          && C->aggressive_unboxing() && (base != NULL) && (base == address->in(AddPNode::Base))
+                          && phase->type(base)->higher_equal(TypePtr::NOTNULL);
+  return base_is_phi && load_boxed_value;
+}
+
+Node* LoadNode::phi_or_self(PhaseGVN* phase) {
+  Node* mem = in(Memory);
+  const TypeOopPtr *t_oop = in(Address)->bottom_type()->isa_oopptr();
+  const Type* this_type = bottom_type();
+
+  if (is_instance_field_load_with_local_phi()) {
+    Node *region = mem->in(0);
+
+    int this_index  = phase->C->get_alias_index(t_oop);
+    int this_offset = t_oop->offset();
+    int this_iid    = t_oop->instance_id();
+    for (DUIterator_Fast imax, i = region->fast_outs(imax); i < imax; i++) {
+      Node* phi = region->fast_out(i);
+      if (phi->is_Phi() && phi != mem &&
+          phi->as_Phi()->is_same_inst_field(this_type, (int)mem->_idx, this_iid, this_index, this_offset)) {
+        return phi;
+      }
+    }
+  }
+
+  if (is_boxed_value_load_with_local_phi(phase)) {
```

Comment:
Does `t_oop->is_ptr_to_boxed_value()` hold here? Should we assert this?

Reply:
Since there is not too much code after `t_oop->is_ptr_to_boxed_value()` is checked `is_boxed_value_load_with_local_phi()`, I think this assert might be unnecessary.
https://github.com/openjdk/jdk/pull/21134/files#diff-ec810a1df7150822244e55ee309c86d6cbffe108ae9c72b6d258ea5758677c28R1276-R1280
```cpp
  bool base_is_phi = (base != NULL) && base->is_Phi();
  bool load_boxed_value = t_oop != NULL && t_oop->is_ptr_to_boxed_value()
                          && C->aggressive_unboxing() && (base != NULL) && (base == address->in(AddPNode::Base))
                          && phase->type(base)->higher_equal(TypePtr::NOTNULL);
  return base_is_phi && load_boxed_value;
  ```